### PR TITLE
Fix for #33: Let useQuery accept a generic type to infer response type

### LINKED ISSUE #33: Make useQuery accept generic type to infer response?
Currently `useQuery`s props object is just of type `Object`.

I am using `relay-compiler-language-typescript` to create typed artifacts, so it would be really awesome if this library could support typing the response of useQuery.

The created artifact typed look something like this (just an example):
```ts
import { ConcreteRequest } from "relay-runtime";
export type IndexQueryVariables = {};
export type IndexQueryResponse = {
    readonly rootCategories: ReadonlyArray<{
        readonly id: string;
        readonly name: string;
    }> | null;
};
export type IndexQuery = {
    readonly response: IndexQueryResponse;
    readonly variables: IndexQueryVariables;
};
```

Inside a component it would be really great if we could so something like this:

```ts
const {props, error} = useQuery<IndexQuery>({
  query,
  variables: {}
});
```

Then the `useQuery` hook could assign `IndexQuery.response` to it's `props` return key.
Is this something you would consider?

### PR DESCRIPTION
**Pull request by JohannesKlauss**
Fixes #33 
This would be my implementation proposal which closely mimics the actual implementation of react-relay.

I am not quite sure if we can rely on types react-relay provides itself? Then we could just use their `OperationType` implementation which would save a few lines.

### CODE REVIEW

**Comment by morrys** [comment]
Tomorrow I try to look better and then integrate the PR, I have doubts about the typing of the variables
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/react-relay/index.d.ts#L119

**Comment by JohannesKlauss** [comment]
But the variables are never returned from the hook, or am I missing something?

**Comment by taion** [comment]
You want to use `T['variables']` to check the type of the variables given to the hook.

**Comment by morrys** [comment]
No the variables are not returned.

But it is an opportunity to fix the typing and I thought to add this:

* useQuery
```
const useQuery = function<TOperationType extends OperationType> (props: UseQueryProps<TOperationType>) {
```

* RelayHooksType
```
export interface Variables {
     [name: string]: any;
}

export type OperationType = {
     readonly response: unknown;
     readonly variables: Variables;
};

export interface RenderProps<T extends OperationType> {
     error: Error,
     props: T ['response'],
     retry: () => void,
     cached ?: boolean
};

export interface UseQueryProps<T extends OperationType> {
     dataFrom ?: DataFrom,
     query: GraphQLTaggedNode,
     variables: T ['variables'],
};
```

**Comment by morrys** [comment]
I didn't check the code because I don't have access to my PC...

I would also remove Reference from useQuery and move it to RelayHooksType

**Comment by morrys** [comment]
If you can't do it, as soon as I can I accept the PR and make these changes :)

**Comment by JohannesKlauss** [comment]
I can do this tomorrow. If you're faster, so be it :)

**Comment by morrys** [comment]
I will accept the PR and release the new version over the weekend.
Other suggestions for improving typing are welcome.

File: src/RelayHooksType.ts [hunk 1]
```diff
@@ -19,23 +16,31 @@ interface DataFromEnum {
 };
 export type DataFrom = keyof DataFromEnum;
 
+export interface Variables {
```
Comment (taion) [inline src/RelayHooksType.ts]: remove this interface; see below

File: src/RelayHooksType.ts [hunk 2]
```diff
@@ -19,23 +16,31 @@ interface DataFromEnum {
...
+    [name: string]: any;
+}
+
 export type ContainerResult = {
     data: { [key: string]: any },
     resolver: FragmentSpecResolver,
 };
 
+export type OperationType = {
```
Comment (taion) [inline src/RelayHooksType.ts]:
```js
import { OperationType } from 'relay-runtime';
```

**Pull request merged**